File: apps/web/src/domain/schemas/intelligence.ts
import { z } from "zod";

export const providerTierSchema = z.enum(["A", "B", "C"]);
export const intelligenceSourceTypeSchema = z.enum([
  "news",
  "reddit",
  "x",
  "telegram",
  "discord",
  "research",
  "exchange",
  "protocol",
  "market",
  "onchain",
  "influencer",
]);
export const contentRoleSchema = z.enum([
  "primary",
  "commentary",
  "repost",
  "rumor",
]);

export const signalContributionSchema = z.object({
  narrative_class: z.string(),
  weight: z.string(),
  confidence_delta: z.string(),
  explain_snippet: z.string(),
});

export const intelligenceItemSchema = z.object({
  id: z.string().uuid(),
  external_id: z.string(),
  dedup_key: z.string(),
  observed_at: z.string(),
  ingested_at: z.string(),
  provider: z.string(),
  provider_tier: providerTierSchema,
  source_type: intelligenceSourceTypeSchema,
  original_source: z.string(),
  content_role: contentRoleSchema,
  url: z.string().nullable().optional(),
  title: z.string(),
  body: z.string(),
  language: z.string(),
  asset_tags: z.array(z.string()),
  entity_tags: z.array(z.string()),
  sentiment_score: z.string().nullable().optional(),
  credibility_score: z.string(),
  novelty_score: z.string(),
  cross_source_confirmation_count: z.number().int(),
  confirming_providers: z.array(z.string()),
  narrative_cluster_id: z.string().nullable().optional(),
  signal_contribution: signalContributionSchema.nullable().optional(),
});

export type IntelligenceItem = z.infer<typeof intelligenceItemSchema>;

export const linkedIntelligenceSchema = z.object({
  intelligence_id: z.string().uuid(),
  contribution_weight: z.string(),
  confidence_delta: z.string(),
  explain_snippet: z.string(),
  item: intelligenceItemSchema,
});

export type LinkedIntelligence = z.infer<typeof linkedIntelligenceSchema>;

export const intelligenceTickerSchema = z.object({
  schema_version: z.string(),
  generated_at: z.string(),
  items: z.array(intelligenceItemSchema),
});

export const intelligenceFeedSchema = z.object({
  schema_version: z.string(),
  generated_at: z.string(),
  total: z.number(),
  items: z.array(intelligenceItemSchema),
});

export type IntelligenceFeed = z.infer<typeof intelligenceFeedSchema>;

export const intelligenceSignalsSchema = z.object({
  schema_version: z.string(),
  intelligence_id: z.string().uuid(),
  signal_ids: z.array(z.string().uuid()),
});

File: src/signals_lab/api/schemas/intelligence.py
"""Intelligence API schemas."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SignalContributionSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    narrative_class: str
    weight: Decimal
    confidence_delta: Decimal
    explain_snippet: str


class IntelligenceItemSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    external_id: str
    dedup_key: str
    observed_at: datetime
    ingested_at: datetime
    provider: str
    provider_tier: str
    source_type: str
    original_source: str
    content_role: str
    url: str | None = None
    title: str
    body: str = ""
    language: str = "en"
    asset_tags: list[str] = Field(default_factory=list)
    entity_tags: list[str] = Field(default_factory=list)
    sentiment_score: Decimal | None = None
    credibility_score: Decimal
    novelty_score: Decimal
    cross_source_confirmation_count: int = 1
    confirming_providers: list[str] = Field(default_factory=list)
    narrative_cluster_id: str | None = None
    signal_contribution: SignalContributionSchema | None = None


class LinkedIntelligenceSchema(BaseModel):
    intelligence_id: UUID
    contribution_weight: Decimal
    confidence_delta: Decimal
    explain_snippet: str
    item: IntelligenceItemSchema


class IntelligenceTickerResponse(BaseModel):
    schema_version: str = "1.0.0"
    generated_at: datetime
    items: list[IntelligenceItemSchema]


class IntelligenceFeedResponse(BaseModel):
    schema_version: str = "1.0.0"
    generated_at: datetime
    total: int
    items: list[IntelligenceItemSchema]


class IntelligenceSignalsResponse(BaseModel):
    schema_version: str = "1.0.0"
    intelligence_id: UUID
    signal_ids: list[UUID]


def intelligence_from_domain(item: Any) -> IntelligenceItemSchema:
    """Map domain IntelligenceItem to API schema."""
    payload = item.model_dump(mode="json")
    return IntelligenceItemSchema.model_validate(payload)
